File: java/SwimmingCompetition/SwimmingCompetitionServer/src/main/java/swimmingcompetition/server/SwimmingCompetitionServicesImpl.java
package swimmingcompetition.server;

import swimmingcompetition.model.Challenge;
import swimmingcompetition.model.Organizer;
import swimmingcompetition.model.Participant;
import swimmingcompetition.model.Registration;
import swimmingcompetition.model.dto.ChallengeDTO;
import swimmingcompetition.model.dto.ParticipantDTO;
import swimmingcompetition.persistence.repository.ChallengeDBRepository;
import swimmingcompetition.persistence.repository.OrganizerDBRepository;
import swimmingcompetition.persistence.repository.ParticipantDBRepository;
import swimmingcompetition.persistence.repository.RegistrationDBRepository;
import swimmingcompetition.services.ISwimmingCompetitionObserver;
import swimmingcompetition.services.ISwimmingCompetitionServices;
import swimmingcompetition.services.SwimmingCompetitionException;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


public class SwimmingCompetitionServicesImpl implements ISwimmingCompetitionServices {

    private ChallengeDBRepository challengeDBRepository;
    private OrganizerDBRepository organizerDBRepository;
    private ParticipantDBRepository participantDBRepository;
    private RegistrationDBRepository registrationDBRepository;
    private Map<Long, ISwimmingCompetitionObserver> loggedOrganizers;

    public SwimmingCompetitionServicesImpl(ChallengeDBRepository challengeDBRepository, OrganizerDBRepository organizerDBRepository, ParticipantDBRepository participantDBRepository, RegistrationDBRepository registrationDBRepository) {
        this.challengeDBRepository = challengeDBRepository;
        this.organizerDBRepository = organizerDBRepository;
        this.participantDBRepository = participantDBRepository;
        this.registrationDBRepository = registrationDBRepository;
        this.loggedOrganizers = new ConcurrentHashMap<>();
    }

    @Override
    public synchronized Organizer findAccount(String username, String password) throws SwimmingCompetitionException {
        Organizer organizer = this.organizerDBRepository.findAccount(username, password);
        if(organizer != null){
            if(loggedOrganizers.get(organizer.getId()) != null){
                throw new SwimmingCompetitionException("Organizer already logged.");
            }
        }
        else{
            throw new SwimmingCompetitionException("Organizer not found.");
        }
        return organizer;
    }

    @Override
    public synchronized void login(Organizer user, ISwimmingCompetitionObserver client) throws SwimmingCompetitionException {
        if(loggedOrganizers.get(user.getId())!=null)
            throw new SwimmingCompetitionException("Organizer already logged in.");
        loggedOrganizers.put(user.getId(), client);
    }

    @Override
    public synchronized ChallengeDTO[] findAllChallenges() throws SwimmingCompetitionException{
        List<ChallengeDTO> result = new ArrayList<>();
        Iterable<Challenge> challenges = this.challengeDBRepository.findAll();
        if(challenges == null){
            throw new SwimmingCompetitionException("Challenges not found");
        }
        for(Challenge challenge : challenges){
            int no = this.getNumberOfParticipantsByChallenge(challenge);
            ChallengeDTO dto = new ChallengeDTO(challenge.getId(), challenge.getDistance(), challenge.getStyle(), no);
            result.add(dto);
        }
        return result.toArray(new ChallengeDTO[result.size()]);
    }

    private synchronized int getNumberOfParticipantsByChallenge(Challenge challenge){
        return this.registrationDBRepository.countByChallenge(challenge);
    }

    @Override
    public synchronized List<ParticipantDTO> findParticipantsByChallenge(Challenge challenge) throws SwimmingCompetitionException {
        List<ParticipantDTO> result = new ArrayList<>();
        List<Participant> participants = this.registrationDBRepository.findParticipantsByChallenge(challenge);
        if(participants == null){
            throw new SwimmingCompetitionException("Participants not found.");
        }
        for(Participant participant : participants){
            List<Challenge> challenges = new ArrayList<>();
            ParticipantDTO dto = new ParticipantDTO(participant.getId(), participant.getName(), participant.getAge(), challenges);
            List<Challenge> challengeList = this.findChallengesByParticipant(participant);
            dto.setChallenges(challengeList);
            result.add(dto);
        }
        return result;
    }

    private synchronized List<Challenge> findChallengesByParticipant(Participant participant) throws SwimmingCompetitionException {
        List<Challenge> challenges = this.registrationDBRepository.findChallengesByParticipant(participant);
        if(challenges == null){
            throw new SwimmingCompetitionException("Challenges not found.");
        }
        return challenges;
    }

    @Override
    public synchronized Participant addParticipant(String name, int age) throws SwimmingCompetitionException {
        Participant participant = new Participant(name, age);
        Participant participant1 = this.participantDBRepository.save(participant);

        return participant1;
    }

    private final int defaultThreadsNo=5;
    private void notifyOrganizators() {
        Iterable<Organizer> organizers = this.organizerDBRepository.findAll();
        Iterable<Challenge> challenges = this.challengeDBRepository.findAll();
        List<ChallengeDTO> challengeDTOS = new ArrayList<>();
        for(Challenge challenge : challenges){
            int no = this.getNumberOfParticipantsByChallenge(challenge);
            ChallengeDTO dto = new ChallengeDTO(challenge.getId(), challenge.getDistance(), challenge.getStyle(), no);
            dto.setNoParticipants(no);
            challengeDTOS.add(dto);
        }
        ExecutorService executor= Executors.newFixedThreadPool(defaultThreadsNo);
        for(Organizer organizer : organizers){
            ISwimmingCompetitionObserver observer = loggedOrganizers.get(organizer.getId());
            if(observer != null){
                executor.execute(() -> {
                        System.out.println("Notifying [" + organizer.getId()+ "]  logged in.");
                        observer.updateTables(challengeDTOS.toArray(new ChallengeDTO[challengeDTOS.size()]));
                });
            }
        }
    }

    @Override
    public synchronized void addParticipantToMoreChallenges(Participant participant, List<Challenge> challenges){
        this.registrationDBRepository.addParticipantToMoreChallenges(participant, challenges);
        notifyOrganizators();
    }

    @Override
    public synchronized void logout(Organizer user, ISwimmingCompetitionObserver client) throws SwimmingCompetitionException {
        ISwimmingCompetitionObserver localClient = loggedOrganizers.remove(user.getId());
        if(localClient == null){
            throw new SwimmingCompetitionException("Client not logged in.");
        }
        // notify
    }
    private synchronized static <T> T[] iterableToArray(Iterable<T> iterable, Class<T> clazz) {
        // Creează o listă pentru a stoca elementele din iterable
        List<T> lista = new ArrayList<>();

        // Adaugă elementele din iterable în listă
        for (T element : iterable) {
            lista.add(element);
        }

        // Creează un array de tipul specificat
        @SuppressWarnings("unchecked")
        T[] array = (T[]) java.lang.reflect.Array.newInstance(clazz, lista.size());

        // Copiază elementele din lista în array
        lista.toArray(array);

        return array;
    }
}

File: csharp/SwimmingCompetitionCS/Server/SwimmingCompetitionServicesImpl.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using model;
using persistence;
using services;

namespace Server
{
    public class SwimmingCompetitionServicesImpl: ISwimmingCompetitionServices
    {
        private ChallengeDBRepository challengeDBRepository;
        private OrganizerDBRepository organizerDBRepository;
        private ParticipantDBRepository participantDBRepository;
        private RegistrationDBRepository registrationDBRepository;
        private IDictionary<long, ISwimmingCompetitionObserver> loggedOrganizers;

        public SwimmingCompetitionServicesImpl(ChallengeDBRepository challengeDbRepository, OrganizerDBRepository organizerDbRepository, ParticipantDBRepository participantDbRepository, RegistrationDBRepository registrationDbRepository)
        {
            challengeDBRepository = challengeDbRepository;
            organizerDBRepository = organizerDbRepository;
            participantDBRepository = participantDbRepository;
            registrationDBRepository = registrationDbRepository;
            loggedOrganizers = new Dictionary<long, ISwimmingCompetitionObserver>();
        }

        public void login(model.Organizer user, ISwimmingCompetitionObserver client)
        {
            if(loggedOrganizers.ContainsKey(user.Id))
                throw new SwimmingCompetitionException("Organizer already logged in.");
            loggedOrganizers[user.Id] =  client;
        }

        public model.Organizer findAccount(string username, string password)
        {
            model.Organizer organizer = this.organizerDBRepository.FindAccount(username, password);
            if(organizer != null){
                if(loggedOrganizers.ContainsKey(organizer.Id)){
                    throw new SwimmingCompetitionException("Organizer already logged.");
                }
            }
            else{
                throw new SwimmingCompetitionException("Organizer not found.");
            }
            return organizer;
        }

        public model.ChallengeDTO[] findAllChallenges()
        {
            List<model.ChallengeDTO> result = new List<model.ChallengeDTO>();
            IEnumerable<model.Challenge> challenges = this.challengeDBRepository.FindAll();
            if(challenges == null){
                throw new SwimmingCompetitionException("Challenges not found");
            }

            foreach (var challenge in challenges)
            {
                int no = this.getNumberOfParticipantsByChallenge(challenge);
                model.ChallengeDTO challengeDto = new model.ChallengeDTO(challenge.Id, challenge.Distance, challenge.Style, no);
                result.Add(challengeDto);
            }
            return result.ToArray();
        }

        private int getNumberOfParticipantsByChallenge(model.Challenge challenge)
        { 
            return this.registrationDBRepository.CountByChallenge(challenge);
        }

        public IList<model.ParticipantDTO> findParticipantsByChallenge(model.Challenge challenge)
        {
            IList<model.ParticipantDTO> result = new List<model.ParticipantDTO>();
            IList<model.Participant> participants = this.registrationDBRepository.FindParticipantsByChallenge(challenge);
            if(participants == null){
                throw new SwimmingCompetitionException("Participants not found.");
            }

            foreach (var participant in participants)
            {
                model.ParticipantDTO participantDto = new model.ParticipantDTO(participant.Id, participant.Name, participant.Age,
                    this.findChallengesByParticipant(participant).ToList());
                result.Add(participantDto);
            }
            return result;
        }

        private IList<model.Challenge> findChallengesByParticipant(model.Participant participant)
        {
            IList<model.Challenge> challenges = this.registrationDBRepository.FindChallengesByParticipant(participant);
            if(challenges == null){
                throw new SwimmingCompetitionException("Challenges not found.");
            }
            return challenges;
        }

        public model.Participant addParticipant(string name, int age)
        {
            model.Participant participant = new model.Participant(name, age);

            model.Participant participant1 = this.participantDBRepository.Save(participant);
            return participant1;
        }

        public void addParticipantToMoreChallenges(model.Participant participant, List<model.Challenge> challenges)
        {
            this.registrationDBRepository.AddParticipantToMoreChallenges(participant, challenges);
            // TO DO
            notifyOrganizators(this.challengeDBRepository.FindAll().ToArray());
        }

        private void notifyOrganizators(model.Challenge[] challenges)
        {
            IEnumerable<model.Organizer> organizers = organizerDBRepository.FindAll();
            List<model.ChallengeDTO> result = new List<model.ChallengeDTO>();
            if(challenges == null){
                throw new SwimmingCompetitionException("Challenges not found");
            }
            foreach (var challenge in challenges)
            {
                int no = this.getNumberOfParticipantsByChallenge(challenge);
                model.ChallengeDTO challengeDto = new model.ChallengeDTO(challenge.Id, challenge.Distance, challenge.Style, no);
                result.Add(challengeDto);
            }
            foreach (var organizer in organizers)
            {
                if (loggedOrganizers.ContainsKey(organizer.Id))
                {
                    ISwimmingCompetitionObserver observer = loggedOrganizers[organizer.Id];
                    Task.Run(() => observer.updateTables(result.ToArray()));
                }
            }
        }

        public void logout(model.Organizer user, ISwimmingCompetitionObserver client)
        {
            ISwimmingCompetitionObserver localClient = loggedOrganizers[user.Id];
            if(localClient == null){
                throw new SwimmingCompetitionException("Client not logged in.");
            }

            loggedOrganizers.Remove(user.Id);
        }
    }
}